newclient/client.go

```diff
@@ -0,0 +1,245 @@
+	return func(c *Client) error {
+		c.httpClient = httpClient
+		return nil
+	}
+}
+
+// The WithToken function allows to specify a JWT
+// token that the client will use to interact with Astarte.
+func WithToken(token string) clientOption {
+	return func(c *Client) error {
+		c.token = token
+		return nil
+	}
+}
+
+// The WithUserAgent function allows to specify the User Agent
+// that the client will use when making http requests.
+func WithUserAgent(userAgent string) clientOption {
+	return func(c *Client) error {
+		c.userAgent = userAgent
+		return nil
+	}
+}
+
+// The WithPrivateKey function allows to specify a realm private key,
+// used internally to generate a valid JWT token to all Astarte APIs with no expiry.
+// The client will use that token to interact with Astarte.
+func WithPrivateKey[T privateKeyProvider](privateKey T) clientOption {
+	return WithPrivateKeyWithTTL(privateKey, 0)
+}
+
+// The WithPrivateKey function allows to specify a realm private key,
+// used internally to generate a valid JWT token to all Astarte APIs with a specified expiry (in seconds).
+// The client will use that token to interact with Astarte.
+func WithPrivateKeyWithTTL[T privateKeyProvider](privateKey T, ttlSeconds int64) clientOption {
+	// Add all types
+	servicesAndClaims := map[misc.AstarteService][]string{
+		misc.AppEngine:       {},
+		misc.Channels:        {},
+		misc.Flow:            {},
+		misc.Housekeeping:    {},
+		misc.Pairing:         {},
+		misc.RealmManagement: {},
+	}
+	return WithPrivateKeyWithClaimsWithTTL(privateKey, servicesAndClaims, 0)
+}
+
+// The WithPrivateKey function allows to specify a realm private key,
+// used internally to generate a valid JWT token with a given set of Astarte claims and
+// a specified expiry (in seconds).
+// The client will use that token to interact with Astarte.
+func WithPrivateKeyWithClaimsWithTTL[T privateKeyProvider](privateKey T, claims map[misc.AstarteService][]string, ttlSeconds int64) clientOption {
+	return func(c *Client) error {
+		// Golang I hate you so much
+		switch k := any(privateKey).(type) {
+		case string:
+			var err error
+			c.token, err = misc.GenerateAstarteJWTFromKeyFile(k, claims, ttlSeconds)
+			return err
+		case []byte:
+			var err error
+			c.token, err = misc.GenerateAstarteJWTFromPEMKey(k, claims, ttlSeconds)
+			return err
+		default:
+			return ErrNoPrivateKeyProvided
+		}
+	}
+}
+
+func validate(c *Client) error {
+	if c.baseURL != nil && (c.appEngineURL != nil || c.realmManagementURL != nil || c.housekeepingURL != nil || c.pairingURL != nil) {
+		return ErrConflictingUrls
+	}
+	if c.baseURL == nil && c.appEngineURL == nil && c.realmManagementURL == nil && c.housekeepingURL == nil && c.pairingURL == nil {
+		return ErrNoUrlsProvided
+	}
+	return nil
+}
+
+func setDefaults(c *Client) *Client {
+	if c.httpClient == nil {
+		c.httpClient = &http.Client{
+			Timeout: time.Second * 30,
+		}
+
+	}
+	if c.userAgent == "" {
+		c.userAgent = "astarte-go"
```

Comment:
Maybe also the version?

Reply:
Actually™, since this package is part of a library (which does not even have a main file), I think it would be better to be agnostic wrt the version. Users of the library (e.g. astartectl) might want to specify their own version in the user agent. This UA should just be a fallback.